src/session/sidebar/mod.rs

```diff
@@ -137,42 +154,86 @@ impl Sidebar {
-        let sorter = gtk::CustomSorter::new(move |obj1, obj2| {
-            let order1 = obj1.downcast_ref::<Chat>().unwrap().order();
-            let order2 = obj2.downcast_ref::<Chat>().unwrap().order();
-
-            order2.cmp(&order1).into()
-        });
-        self_.sorter.set(sorter).unwrap();
-
-        let filter = self_.filter.get().unwrap();
-        let sorter = self_.sorter.get().unwrap();
-        chat_list.connect_positions_changed(clone!(@weak filter, @weak sorter => move |_| {
-            filter.changed(gtk::FilterChange::Different);
-            sorter.changed(gtk::SorterChange::Different);
-        }));
-
-        let filter_model = gtk::FilterListModel::new(Some(&chat_list), Some(filter));
-        let sort_model = gtk::SortListModel::new(Some(&filter_model), Some(sorter));
-        let selection = gtk::SingleSelection::new(Some(&sort_model));
-        selection.set_autoselect(false);
-        selection
-            .bind_property("selected-item", self, "selected-chat")
-            .flags(glib::BindingFlags::SYNC_CREATE)
-            .build();
-        self_.selection.replace(Some(selection));
-
-        self_
-            .chat_list_view
-            .set_model(Some(self_.selection.borrow().as_ref().unwrap()));
+        self_.search_bar.set_search_mode(true);
+        self_.search_entry.grab_focus();
+    }
+
+    fn search(&self, query: String) {
+        let self_ = imp::Sidebar::from_instance(self);
+        self_.searched_chats.borrow_mut().clear();
+        self_.searched_users.borrow_mut().clear();
+        self_.already_searched_users.borrow_mut().clear();
+
+        if query.is_empty() {
+            if let Some(filter) = self_.filter.borrow().as_ref() {
+                filter.changed(gtk::FilterChange::Different);
+            }
+        } else {
+            let client_id = self
+                .session()
+                .expect("The session needs to be set to be able to search")
+                .client_id();
+
+            // Search chats
+            let query_clone = query.clone();
+            do_async(
+                glib::PRIORITY_DEFAULT_IDLE,
+                async move {
+                    functions::SearchChats::new()
+                        .query(query_clone)
+                        .limit(100)
+                        .send(client_id)
+                        .await
+                },
+                clone!(@weak self as obj => move |result| async move {
+                    if let Ok(enums::Chats::Chats(chats)) = result {
+                        let self_ = imp::Sidebar::from_instance(&obj);
+
+                        if let Some(filter) = self_.filter.borrow().as_ref() {
+                            let session = obj
+                                .session()
+                                .expect("The session needs to be set to be able to search");
+                            let chat_list = session.chat_list();
+
+                            for chat_id in &chats.chat_ids {
```

Comment:
I am a huge fan of iterators and combinators. Makes everything more functional. But this is personal taste. :-)

```rust
                            self_.already_searched_users.borrow_mut().extend(chats.chat_ids.iter()
                                .filter_map(|id| chat_list.get_chat(*id))
                                .filter_map(|chat| match chat.type_() {
                                    enums::ChatType::Private(types::ChatTypePrivate { user_id }) => Some(*user_id),
                                    _ => None
                                }
                            ));
```